File: modules/core/src/main/java/fr/rozanc/cocozzelle/datatable/DefaultColumnExpressionParser.java
package fr.rozanc.cocozzelle.datatable;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class DefaultColumnExpressionParser implements ColumnExpressionParser {

    private final Pattern modifierPattern = Pattern.compile("\\s*\\((([^),\\s]+)(?:,\\s*[^),\\s]+)*)\\)\\s*$");

    @Override
    public ColumnExpression parse(final String expression) {
        final Matcher matcher = modifierPattern.matcher(expression);
        final Map<String, String> options = new LinkedHashMap<>();
        if (matcher.find()) {
            final String allModifiersString = matcher.group(1);
            final String[] modifiersStrings = allModifiersString.split(",\\s*");
            for (final String modifiersString : modifiersStrings) {
                final String[] elements = modifiersString.split(":", 1);
                if (elements.length > 1) {
                    options.put(elements[0], elements[1]);
                } else {
                    options.put(elements[0], "true");
                }
            }
        }
        return new DefaultColumnExpression(expression,
                                           expression.replaceFirst(modifierPattern.pattern(), ""),
                                           options);
    }
}

File: modules/core/src/main/kotlin/fr/rozanc/kokozzelle/datatable/DefaultColumnExpressionParser.kt
package fr.rozanc.kokozzelle.datatable

import java.util.LinkedHashMap

class DefaultColumnExpressionParser : ColumnExpressionParser {

    companion object {
        private val MODIFIERS_PATTERN = Regex("\\s*\\((([^),\\s]+)(?:,\\s*[^),\\s]+)*)\\)\\s*$")
        private val MODIFIERS_SEPARATOR_PATTERN = Regex(",\\s*")
    }

    override fun parse(expression: String): ColumnExpression {
        val matcher = MODIFIERS_PATTERN.find(expression)
        val options: MutableMap<String, String> = LinkedHashMap()
        if (matcher != null) {
            val allModifiersString = matcher.groups[1]!!.value
            val modifiersStrings = allModifiersString.split(MODIFIERS_SEPARATOR_PATTERN).toTypedArray()
            for (modifiersString in modifiersStrings) {
                val elements = modifiersString.split(":", limit = 1).toTypedArray()
                if (elements.size > 1) {
                    options[elements[0]] = elements[1]
                } else {
                    options[elements[0]] = "true"
                }
            }
        }
        return DefaultColumnExpression(
            expression,
            expression.replaceFirst(MODIFIERS_PATTERN, "").trim(),
            options
        )
    }


}
